README.md
# dooz_color_picker

A Flutter package that provides widgets to enable a user to select a color and some other params of a RGB light using circular sliders.

## Getting Started
```dart
import 'package:dooz_color_picker/dooz_color_picker.dart';

CircleColorPicker(
  radius: 140,
  thumbRadius: 15,
  initialColor: Colors.red,
  child: Text(pickerColor?.toString() ?? ''),
  colorListener: (Color value) {
    setState(() {
      pickerColor = value;
    });
 },
)

CircleTemperaturePicker(
  radius: 140,
  thumbRadius: 15,
  initialTemperature: 2600,
  startTemperature: 2600,
  endTemperature: 10600,
  colorListener: (Color value, int kDegree) {
    setState(() {
      kelvinColor = value;
      kelvinValue = kDegree;
    });
  },
  child: Text(
    '${kelvinColor?.toString() ?? ''}\n'
    '${kelvinValue?.toString() ?? ''}',
    textAlign: TextAlign.center,
  ),
),
```
![example sliders](./example.png)


CHANGELOG.md
## 0.1.0

- initial open source release


lib/src/color_picker.dart
import 'dart:math';

import 'package:flutter/material.dart';

/// A listener which receives an color in int representation. as used
/// by [CircleColorPicker.colorListener].
typedef ColorListener = void Function(Color value);

/// A circle palette color picker.
class CircleColorPicker extends StatefulWidget {
  // radius of the color palette, note that radius * 2 is not the final
  // width of this widget, instead is (radius + thumbRadius) * 2.
  final double radius;

  /// radius of thumb.
  final double thumbRadius;

  /// thumb stroke color.
  final Color thumbStrokeColor;

  /// A listener receives color pick events.
  final ColorListener colorListener;

  /// initial color of this color picker.
  final Color initialColor;

  /// Child widget
  final Widget? child;

  const CircleColorPicker({
    Key? key,
    this.radius = 160,
    this.initialColor = const Color(0xffff0000),
    this.thumbStrokeColor = Colors.white,
    this.thumbRadius = 8,
    this.child,
    required this.colorListener,
  }) : super(key: key);

  @override
  State<CircleColorPicker> createState() {
    return _CircleColorPickerState();
  }
}

class _CircleColorPickerState extends State<CircleColorPicker> {
  static const List<Color> colors = [
    Color(0xffff0000),
    Color(0xffff4000),
    Color(0xffff8000),
    Color(0xffffbf00),
    Color(0xffffff00),
    Color(0xffbfff00),
    Color(0xff80ff00),
    Color(0xff40ff00),
    Color(0xff00ff00),
    Color(0xff00ff40),
    Color(0xff00ff80),
    Color(0xff00ffbf),
    Color(0xff00ffff),
    Color(0xff00bfff),
    Color(0xff0080ff),
    Color(0xff0040ff),
    Color(0xff0000ff),
    Color(0xff4000ff),
    Color(0xff8000ff),
    Color(0xffbf00ff),
    Color(0xffff00ff),
    Color(0xffff00bf),
    Color(0xffff0080),
    Color(0xffff0040),
  ];

  late double thumbDistanceToCenter;
  late double thumbRadians;
  late Color color;

  @override
  void initState() {
    super.initState();
    thumbDistanceToCenter = widget.radius - widget.thumbRadius;
    final double hue = HSVColor.fromColor(widget.initialColor).hue;
    thumbRadians = degreesToRadians(270 - hue);
    color = widget.initialColor;
  }

  @override
  Widget build(BuildContext context) {
    final double radius = widget.radius;
    final double thumbRadius = widget.thumbRadius;

    // compute thumb center coordinate
    final double thumbCenterX =
        radius + thumbDistanceToCenter * sin(thumbRadians);
    final double thumbCenterY =
        radius + thumbDistanceToCenter * cos(thumbRadians);

    // build thumb widget
    Widget thumb = Positioned(
        left: thumbCenterX,
        top: thumbCenterY,
        child: Container(
          width: thumbRadius * 2,
          height: thumbRadius * 2,
          decoration: BoxDecoration(
              boxShadow: [
                BoxShadow(
                  color: Colors.black.withOpacity(0.2),
                  offset: const Offset(0, 2),
                  blurRadius: 10,
                  spreadRadius: 1,
                )
              ],
              border: Border.all(
                  color: widget.thumbStrokeColor, width: thumbRadius * 0.25),
              color: color,
              borderRadius: BorderRadius.all(Radius.circular(thumbRadius))),
        ));
    return GestureDetector(
        behavior: HitTestBehavior.opaque,
        onPanDown: (details) => handleTouch(details.globalPosition, context),
        onPanStart: (details) => handleTouch(details.globalPosition, context),
        onPanUpdate: (details) => handleTouch(details.globalPosition, context),
        child: Stack(
          children: <Widget>[
            SizedBox(
                width: (radius + thumbRadius) * 2,
                height: (radius + thumbRadius) * 2),
            Positioned(
              left: thumbRadius,
              top: thumbRadius,
              child: Container(
                width: radius * 2,
                height: radius * 2,
                decoration: BoxDecoration(
                    borderRadius: BorderRadius.all(Radius.circular(radius)),
                    gradient: const SweepGradient(colors: colors)),
              ),
            ),
            Positioned(
              left: thumbRadius * 3,
              top: thumbRadius * 3,
              child: Container(
                width: (radius - thumbRadius * 2) * 2,
                height: (radius - thumbRadius * 2) * 2,
                decoration: BoxDecoration(
                  color: Colors.white,
                  borderRadius: BorderRadius.all(
                    Radius.circular(radius - thumbRadius * 2),
                  ),
                ),
                child: Center(child: widget.child ?? Container()),
              ),
            ),
            thumb
          ],
        ));
  }

  /// calculate colors picked from palette and update our states.
  void handleTouch(Offset globalPosition, BuildContext context) {
    RenderBox box = context.findRenderObject() as RenderBox;
    Offset localPosition = box.globalToLocal(globalPosition);
    final double centerX = box.size.width / 2;
    final double centerY = box.size.height / 2;
    final double deltaX = localPosition.dx - centerX;
    final double deltaY = localPosition.dy - centerY;
    double theta = atan2(deltaX, deltaY);
    double degree = 270 - radiansToDegrees(theta);
    if (degree < 0) degree = 360 + degree;
    widget.colorListener(HSVColor.fromAHSV(1, degree, 1, 1).toColor());
    setState(() {
      thumbDistanceToCenter = widget.radius - widget.thumbRadius;
      thumbRadians = theta;
      color = HSVColor.fromAHSV(1, degree, 1, 1).toColor();
    });
  }

  /// convert an angle value from radian to degree representation.
  double radiansToDegrees(double radians) {
    return (radians + pi) / pi * 180;
  }

  /// convert an angle value from degree to radian representation.
  double degreesToRadians(double degrees) {
    return degrees / 180 * pi - pi;
  }
}


lib/src/custom_painter.dart
import 'dart:math';
import 'package:flutter/material.dart';

class CircleProgressBarPainter extends CustomPainter {
  final double strokeWidth;
  final List<Color> foregroundColors;

  CircleProgressBarPainter({
    required this.foregroundColors,
    double? strokeWidth,
  }) : strokeWidth = strokeWidth ?? 6;

  @override
  void paint(Canvas canvas, Size size) {
    final Offset center = size.center(Offset.zero);
    final Size constrainedSize =
        Size(size.width - strokeWidth, size.height - strokeWidth);
    final shortestSide = min(constrainedSize.width, constrainedSize.height);
    final radius = (shortestSide / 2);
    Rect rect = Rect.fromCircle(center: center, radius: radius);
    final s = SweepGradient(
      colors: foregroundColors,
      startAngle: pi / 2 - pi / 4,
      endAngle: pi * 2 - pi / 4,
      transform: const GradientRotation(pi / 2),
    );
    final foregroundPaint = Paint()
      ..shader = s.createShader(rect)
      ..strokeWidth = strokeWidth
      ..strokeCap = StrokeCap.round
      ..style = PaintingStyle.stroke;

    // Start at the top. 0 radians represents the right edge
    const double startAngle = pi / 2 - pi / 4 + pi / 2;
    const double sweepAngle = pi * 2 - pi / 2;

    canvas.drawArc(
      Rect.fromCircle(center: center, radius: radius),
      startAngle,
      sweepAngle,
      false,
      foregroundPaint,
    );
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    final oldPainter = (oldDelegate as CircleProgressBarPainter);
    return oldPainter.foregroundColors != foregroundColors ||
        oldPainter.strokeWidth != strokeWidth;
  }
}


lib/src/gradient_slider.dart
import 'dart:math';
import 'package:flutter/material.dart';

typedef GradientSliderOnChange = void Function(int value);

/// A bar color picker
class GradientSlider extends StatefulWidget {
  /// background Color
  final List<Color> colors;

  /// width of bar
  final double width;

  /// width of bar
  final double height;

  /// A listener receives value (0 => 100).
  final GradientSliderOnChange onChange;

  /// thumb fill color
  final Color thumbColor;

  /// radius of thumb
  final double thumbWidth;

  /// init value
  final int initValue;

  const GradientSlider({
    Key? key,
    this.colors = const [Colors.white, Colors.black],
    this.width = 200,
    this.height = 16,
    this.thumbWidth = 20,
    this.thumbColor = Colors.white,
    this.initValue = 50,
    required this.onChange,
  })  : assert(initValue >= 0 && initValue <= 100),
        super(key: key);

  @override
  _GradientSliderState createState() => _GradientSliderState();
}

class _GradientSliderState extends State<GradientSlider> {
  double percent = 0.0;

  @override
  void initState() {
    percent = widget.initValue / 100;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    double thumbLeft;
    thumbLeft = (widget.width - widget.thumbWidth) * percent;

    Widget frame = SizedBox(width: widget.width, height: widget.thumbWidth);

    // build thumb
    Widget thumb = Positioned(
        left: thumbLeft,
        child: Container(
          width: widget.thumbWidth,
          height: widget.thumbWidth,
          decoration: BoxDecoration(
              boxShadow: [
                BoxShadow(
                  color: Colors.black.withOpacity(0.5),
                  offset: const Offset(0, 2),
                  blurRadius: 10,
                  spreadRadius: 1,
                )
              ],
              color: widget.thumbColor,
              borderRadius:
                  BorderRadius.all(Radius.circular(widget.thumbWidth / 2))),
        ));
    Widget content = Positioned(
      left: 0,
      top: (widget.thumbWidth - widget.height) / 2,
      child: Container(
        width: widget.width,
        height: widget.height,
        decoration: BoxDecoration(
            boxShadow: [
              BoxShadow(
                color: Colors.black.withOpacity(0.2),
                offset: const Offset(0, 2),
                blurRadius: 10,
                spreadRadius: 1,
              )
            ],
            borderRadius: BorderRadius.all(Radius.circular(widget.height / 2)),
            gradient: LinearGradient(colors: widget.colors)),
      ),
    );

    return GestureDetector(
      onPanDown: (details) => handleTouch(details.globalPosition, context),
      onPanStart: (details) => handleTouch(details.globalPosition, context),
      onPanUpdate: (details) => handleTouch(details.globalPosition, context),
      child: Stack(
        children: [frame, content, thumb],
      ),
    );
  }

  /// calculate colors picked from palette and update our states.
  void handleTouch(Offset globalPosition, BuildContext context) {
    RenderBox box = context.findRenderObject() as RenderBox;
    Offset localPosition = box.globalToLocal(globalPosition);
    double percent;
    percent = localPosition.dx / (widget.width);
    percent = min(max(0.0, percent), 1.0);
    setState(() {
      this.percent = percent;
    });
    widget.onChange((percent * 100).toInt());
  }
}


lib/src/color_temperature_picker.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'custom_painter.dart';

/// A listener which receives an color in int representation. as used
/// by [CircleColorPicker.colorListener].
typedef ColorTemperatureListener = void Function(Color color, int kDegree);

/// A circle palette color picker.
class CircleTemperaturePicker extends StatefulWidget {
  // radius of the color palette, note that radius * 2 is not the final
  // width of this widget, instead is (radius + thumbRadius) * 2.
  final double radius;

  /// radius of thumb.
  final double thumbRadius;

  /// thumb stroke color.
  final Color thumbStrokeColor;

  /// A listener receives color pick events.
  final ColorTemperatureListener colorListener;

  /// Start Temperature (geater than 1500)
  final int startTemperature;

  /// End Temperature (less than 40000)
  final int endTemperature;

  /// initial Temperature of this picker.
  final int initialTemperature;

  /// Child widget
  final Widget? child;

  const CircleTemperaturePicker(
      {Key? key,
      this.radius = 160,
      this.thumbRadius = 8,
      this.initialTemperature = 2600,
      this.thumbStrokeColor = Colors.white,
      this.startTemperature = 2600,
      this.endTemperature = 10600,
      this.child,
      required this.colorListener})
      : assert(startTemperature > 1500),
        assert(endTemperature < 40000),
        assert(initialTemperature >= startTemperature &&
            initialTemperature <= endTemperature),
        super(key: key);

  @override
  State<CircleTemperaturePicker> createState() {
    return _TemperaturePickerState();
  }
}

class _TemperaturePickerState extends State<CircleTemperaturePicker> {
  List<Color> colors = [];

  late double thumbDistanceToCenter;
  late double thumbRadians;
  late int colorIndex;

  @override
  void initState() {
    super.initState();
    colors = generateColorsFormKelvin(
        widget.startTemperature, widget.endTemperature);
    thumbDistanceToCenter = widget.radius - widget.thumbRadius;
    int initColor = colors.indexWhere(
        (element) => element == getColorFromKelvin(widget.initialTemperature));
    if (initColor == colors.length - 1) initColor = colors.length;
    final degree = 180 - ((270 / colors.length * initColor)) - 45;
    colorIndex = initColor;
    thumbRadians = degreesToRadians(degree);
  }

  @override
  Widget build(BuildContext context) {
    final double radius = widget.radius;
    final double thumbRadius = widget.thumbRadius;

    // compute thumb center coordinate
    final double thumbCenterX =
        radius + thumbDistanceToCenter * sin(thumbRadians);
    final double thumbCenterY =
        radius + thumbDistanceToCenter * cos(thumbRadians);

    // build thumb widget
    Widget thumb = Positioned(
        left: thumbCenterX,
        top: thumbCenterY,
        child: Container(
          width: thumbRadius * 2,
          height: thumbRadius * 2,
          decoration: BoxDecoration(
              boxShadow: [
                BoxShadow(
                  color: Colors.black.withOpacity(0.2),
                  offset: const Offset(0, 2),
                  blurRadius: 10,
                  spreadRadius: 1,
                )
              ],
              border: Border.all(
                  color: widget.thumbStrokeColor, width: thumbRadius * 0.25),
              color: colors[colorIndex],
              borderRadius: BorderRadius.all(Radius.circular(thumbRadius))),
        ));
    return GestureDetector(
        behavior: HitTestBehavior.opaque,
        onPanDown: (details) => handleTouch(details.globalPosition, context),
        onPanStart: (details) => handleTouch(details.globalPosition, context),
        onPanUpdate: (details) => handleTouch(details.globalPosition, context),
        child: Stack(
          children: <Widget>[
            SizedBox(
                width: (radius + thumbRadius) * 2,
                height: (radius + thumbRadius) * 2),
            Positioned(
              left: thumbRadius * 3,
              top: thumbRadius * 3,
              child: Container(
                width: (radius - thumbRadius * 2) * 2,
                height: (radius - thumbRadius * 2) * 2,
                decoration: BoxDecoration(
                  color: Colors.white,
                  borderRadius: BorderRadius.all(
                    Radius.circular(radius - thumbRadius * 2),
                  ),
                ),
                child: Center(child: widget.child ?? Container()),
              ),
            ),
            Positioned(
              left: thumbRadius,
              top: thumbRadius,
              // ignore: sized_box_for_whitespace
              child: Container(
                width: radius * 2,
                height: radius * 2,
                child: CustomPaint(
                  foregroundPainter: CircleProgressBarPainter(
                    foregroundColors: colors,
                    strokeWidth: thumbRadius * 2,
                  ),
                ),
              ),
            ),
            thumb
          ],
        ));
  }

  /// calculate colors picked from palette and update our states.
  void handleTouch(Offset globalPosition, BuildContext context) {
    RenderBox box = context.findRenderObject() as RenderBox;
    Offset localPosition = box.globalToLocal(globalPosition);
    final double centerX = box.size.width / 2;
    final double centerY = box.size.height / 2;
    final double deltaX = localPosition.dx - centerX;
    final double deltaY = localPosition.dy - centerY;
    double theta = atan2(deltaX, deltaY);
    double degree = (180 - radiansToDegrees(theta)).roundToDouble();
    final degreePindex = 270 / colors.length;
    if (degree < 0) degree = 360 + degree;
    if (degree <= 44) return;
    if (degree >= 316) return;

    int index = ((degree - 45) ~/ degreePindex).toInt();
    if (index < 0) index = 0;
    if (index >= colors.length) index = colors.length - 1;
    final kDegree = widget.startTemperature + 100 * colorIndex;
    widget.colorListener(colors[index.toInt()], kDegree);
    setState(() {
      thumbDistanceToCenter = widget.radius - widget.thumbRadius;
      thumbRadians = theta;
      colorIndex = index.toInt();
    });
  }

  /// convert an angle value from radian to degree representation.
  double radiansToDegrees(double radians) {
    return (radians + pi) / pi * 180;
  }

  /// convert an angle value from degree to radian representation.
  double degreesToRadians(double degrees) {
    return degrees / 180 * pi - pi;
  }

  List<Color> generateColorsFormKelvin(int start, int end) {
    List<Color> returnValue = [];
    int flag = start;
    while (flag <= end) {
      final tempColor = getColorFromKelvin(flag);
      returnValue.add(tempColor);
      flag += 100;
    }
    return returnValue;
  }
}

Color getColorFromKelvin(int k) {
  double tmpCalc;
  double tmpKelvin = k.toDouble();
  if (tmpKelvin < 1000) tmpKelvin = 1000;
  if (tmpKelvin > 40000) tmpKelvin = 40000;
  tmpKelvin = tmpKelvin / 100;

  // Red
  double r;
  if (tmpKelvin <= 66) {
    r = 255;
  } else {
    tmpCalc = tmpKelvin - 60;
    tmpCalc = 329.698727446 * pow(tmpCalc, -0.1332047592);
    r = tmpCalc;
    if (r < 0) r = 0;
    if (r > 255) r = 255;
  }
  // Green
  double g;
  if (tmpKelvin <= 66) {
    tmpCalc = tmpKelvin;
    tmpCalc = 99.4708025861 * log(tmpCalc) - 161.1195681661;
    g = tmpCalc;
    if (g < 0) g = 0;
    if (g > 255) g = 255;
  } else {
    tmpCalc = tmpKelvin - 60;
    tmpCalc = 288.1221695283 * pow(tmpCalc, -0.0755148492);
    g = tmpCalc;
    if (g < 0) g = 0;
    if (g > 255) g = 255;
  }
// Blue
  double b;
  if (tmpKelvin >= 66) {
    b = 255;
  } else if (tmpKelvin <= 19) {
    b = 0;
  } else {
    tmpCalc = tmpKelvin - 10;
    tmpCalc = 138.5177312231 * log(tmpCalc) - 305.0447927307;
    b = tmpCalc;
    if (b < 0) b = 0;
    if (b > 255) b = 255;
  }
  return Color.fromRGBO(r.toInt(), g.toInt(), b.toInt(), 1);
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:dooz_color_picker/dooz_color_picker.dart';

void main() => runApp(const ColorPickerApp());

class ColorPickerApp extends StatefulWidget {
  const ColorPickerApp({Key? key}) : super(key: key);

  @override
  _ColorPickerAppState createState() => _ColorPickerAppState();
}

class _ColorPickerAppState extends State<ColorPickerApp> {
  Color? pickerColor;

  //
  Color? kelvinColor;
  int? kelvinValue;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "ColorPicker",
      home: Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
          title: const Text("ColorPicker"),
        ),
        body: Builder(
          builder: (c) {
            return Center(
              child: FittedBox(
                fit: BoxFit.scaleDown,
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: <Widget>[
                    CircleColorPicker(
                      thumbRadius: 15,
                      radius: MediaQuery.of(c).size.width / 2 - 40,
                      initialColor: Colors.red,
                      child: Text(pickerColor?.toString() ?? ''),
                      colorListener: (Color value) {
                        setState(() {
                          pickerColor = value;
                        });
                      },
                    ),
                    CircleTemperaturePicker(
                      thumbRadius: 15,
                      radius: MediaQuery.of(c).size.width / 2 - 40,
                      initialTemperature: 2600,
                      startTemperature: 2600,
                      endTemperature: 10600,
                      colorListener: (Color value, int kDegree) {
                        setState(() {
                          kelvinColor = value;
                          kelvinValue = kDegree;
                        });
                      },
                      child: Text(
                        '${kelvinColor?.toString() ?? ''}\n'
                        '${kelvinValue?.toString() ?? ''}',
                        textAlign: TextAlign.center,
                      ),
                    ),
                  ],
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}
